### ISSUE TITLE
Get-ADDbaccount padding is invalid and cannot be removed

### ISSUE THREAD

**Issue by Priya1964** [MichaelGrafnetter/DSInternals#76]

**Comment by Priya1964** [comment]
Hi Michael,

I am trying the offline module and getting the error of 

Get-ADDbaccount padding is invalid and cannot be removed Can you help with this?

**Comment by MichaelGrafnetter** [comment]
Hi @Priya1964 , thanks for reporting. What version v DSInternals are you using? And from what server does the DB originate?

**Comment by Priya1964** [comment]
I am using version 3.0. I am using an IFM so ntds.dit is on the same server

**Comment by Priya1964** [comment]
Yes it is windows 2016 let me upgrade and try this

**Comment by Priya1964** [comment]
Here You go
Get-ADDBAccount : Padding is invalid and cannot be removed.
At line:1 char:1
+ Get-ADDBAccount -all -DBPath '<C:\ifm\ntds.dit'-BootKe> ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Get-ADDBAccount], CryptographicException
    + FullyQualifiedErrorId : System.Security.Cryptography.CryptographicException,DSInternals.PowerShell.Commands.Ge 
   tADDBAccountCommand


[‎2/‎5/‎2019 8:59 PM]  x2sachdevp:  
No Title 
   at System.Security.Cryptography.CapiSymmetricAlgorithm.DepadBlock(Byte[] block, Int32 offset, Int32 count)
   at System.Security.Cryptography.CapiSymmetricAlgorithm.TransformFinalBlock(Byte[] inputBuffer, Int32 inputOffset, Int32 inputCount)
   at System.Security.Cryptography.CryptoStream.FlushFinalBlock()
   at System.Security.Cryptography.CryptoStream.Dispose(Boolean disposing)
   at System.IO.Stream.Close()
   at DSInternals.Common.Cryptography.DirectorySecretDecryptor.DecryptUsingAES(Byte[] data, Byte[] iv, Byte[] key)
   at DSInternals.DataStore.DataStoreSecretDecryptor.DecryptPekList(Byte[] encryptedBlob, Byte[] bootKey) in D:\a\1\s\Src\DSInternals.DataStore\Cryptography\DataStoreSecretDecryptor.cs:line 341
   at DSInternals.DataStore.DirectoryAgent.GetSecretDecryptor(Byte[] bootKey) in D:\a\1\s\Src\DSInternals.DataStore\DirectoryAgent.PasswordManagement.cs:line 289
   at DSInternals.DataStore.DirectoryAgent.<GetAccounts>d__9.MoveNext() in D:\a\1\s\Src\DSInternals.DataStore\DirectoryAgent.cs:line 63
   at DSInternals.PowerShell.Commands.GetADDBAccountCommand.ReturnAllAccounts(Byte[] bootKey)
   at System.Management.Automation.CommandProcessor.ProcessRecord()

**Comment by c3rberus** [comment]
I am running into the same issue, using version 3.2.1 on Windows 2016.

```
$key = Get-BootKey -SystemHivePath 'C:\IT\troyhunt\hklm_system'
Get-ADDBAccount -All -DBPath 'C:\IT\troyhunt\ntds.dit' -BootKey $key | Format-Custom -View HashcatNT | Out-File C:\IT\users -Encoding ASCII
```
This throws an error...

```
Get-ADDBAccount : Padding is invalid and cannot be removed.
At line:1 char:1
+ Get-ADDBAccount -All -DBPath C:\IT\troyhunt\ntds.dit -BootKey $key |  ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Get-ADDBAccount], CryptographicException
    + FullyQualifiedErrorId : System.Security.Cryptography.CryptographicException,DSInternals.PowerShell.Commands.GetA
   DDBAccountCommand
```
The trace is below...

`
PS C:\IT\troyhunt> $Error[0].Exception.StackTrace
   at System.Security.Cryptography.CapiSymmetricAlgorithm.DepadBlock(Byte[] block, Int32 offset, Int32 count)
   at System.Security.Cryptography.CapiSymmetricAlgorithm.TransformFinalBlock(Byte[] inputBuffer, Int32 inputOffset, Int32 inputCount)
   at System.Security.Cryptography.CryptoStream.FlushFinalBlock()
   at System.Security.Cryptography.CryptoStream.Dispose(Boolean disposing)
   at System.IO.Stream.Close()
   at DSInternals.Common.Cryptography.DirectorySecretDecryptor.DecryptUsingAES(Byte[] data, Byte[] iv, Byte[] key)
   at DSInternals.DataStore.DataStoreSecretDecryptor.DecryptPekList(Byte[] encryptedBlob, Byte[] bootKey) in D:\a\1\s\Src\DSInternals.DataStore\Cryptography\DataStoreSecretDecryptor.cs:line 341
   at DSInternals.DataStore.DirectoryAgent.GetSecretDecryptor(Byte[] bootKey) in D:\a\1\s\Src\DSInternals.DataStore\DirectoryAgent.PasswordManagement.cs:line 289
   at DSInternals.DataStore.DirectoryAgent.<GetAccounts>d__9.MoveNext() in D:\a\1\s\Src\DSInternals.DataStore\DirectoryAgent.cs:line 63
   at DSInternals.PowerShell.Commands.GetADDBAccountCommand.ReturnAllAccounts(Byte[] bootKey)
   at System.Management.Automation.CommandProcessor.ProcessRecord()
`

**Comment by c3rberus** [comment]
I tried an older version of DSInternals (2.23), it then threw the **Get-ADDBaccount: Invalid PEK signature** error. I double checked my steps and sure enough I had the wrong HKLM\System. Once I grabbed the correct one running Get-ADDBAccount worked on 2.23 and 3.2.1.

**Comment by Priya1964** [comment]
It doesnt work for me Though even with 2.23 version

**Comment by MichaelGrafnetter** [comment]
Thanks, @Priya1964 . I have not been able to replicate the issue in my test lab yet. The problem is that the data structures used are totally undocumented. I will have to create a new version of DSInternals which will add the actual value of the [pekList](https://docs.microsoft.com/en-us/windows/desktop/adschema/a-peklist) attribute to the exception.

**Comment by Priya1964** [comment]
Thanks Michael for your quick response. Could you please tell me if Test-passwordquality can be run against a file instead of piping the output from get-adreplaccount which is pulling all accounts information I want to do this for limited accounts so I dumped the hashes in a file and then run it with Test-passwordquality

**Comment by MichaelGrafnetter** [comment]
Have you tried Export-Clixml?

Michael Grafnetter

________________________________

**Comment by Priya1964** [comment]
yes I tried that and it works thanks for that.
However I am seeing a weird issue which is that if we run the module against the domain it yields nothing and in the section of weak passowrds in dictionary and somehow historical passwords section is missing.

**Comment by MichaelGrafnetter** [comment]
Hi @Priya1964, sorry for the delay. I have just published DSInternals 3.3. When the "Padding is invalid and cannot be removed." problem occurs, the resulting exception now has a property called `Data`, and it should contain values `encryptedPEKListBlob` and `bootKey`. I wonder if you could please send me these values so I can fix this issue. Don't worry, they do not contain any identifier of your AD domain and their knowledge does not pose any security risk.

Regarding `Test-PasswordQuality`, could you please open another issue for that? Just note that if you wish to launch it against a subset of accounts, you could use the `Where-Object` cmdlet to filter them out.

**Comment by MichaelGrafnetter** [comment]
Fixed, see issue #91.
